### PR TITLE
core(crc): exclude non network nodes from being a leaf

### PR DESCRIPTION
**Pull request by paulirish**
I ran LH on nytimes and noticed data uris in the crc graph:
image

They seem out of place.

To me, CRC is about optimizing network load, but datauris are inlined into the parent request, so I dont think it's fair to report them in here.

### CODE REVIEW

File: lighthouse-core/computed/critical-request-chains.js [hunk 1]
```diff
@@ -55,6 +56,9 @@ class CriticalRequestChains {
       return false;
     }
 
+    // It's not a request loaded over the network, it's not "critical to load"; it arrived already.
```
Comment (patrickhulce) [inline lighthouse-core/computed/critical-request-chains.js]:
```suggestion
    // If it's not a request loaded over the network, it's not "critical to load"; it arrived already.
```

Comment (patrickhulce) [reply lighthouse-core/computed/critical-request-chains.js]: while we're improving it :)

File: lighthouse-core/test/computed/critical-request-chains-test.js [hunk 2]
```diff
@@ -299,21 +299,39 @@ describe('CriticalRequestChain gatherer: extractChain function', () => {
...
+      0: {
+        request: networkRecords[0],
+        children: {},
+      },
+    });
+  });
+
+
+  it('discards datauris as non-critical', () => {
+    const networkRecords = mockTracingData([HIGH, HIGH, HIGH], [[0, 1], [0, 2]]);
```
Comment (brendankenny) [inline lighthouse-core/test/computed/critical-request-chains-test.js]: `network-records-to-devtools-log.js` does this a lot better :)

File: lighthouse-core/audits/uses-rel-preload.js [hunk 3]
```diff
@@ -104,8 +104,6 @@ class UsesRelPreloadAudit extends Audit {
     if (request.isLinkPreload) return false;
     // It's not critical, don't recommend it.
     if (!CriticalRequestChains.isCritical(request, mainResource)) return false;
```
Comment (brendankenny) [inline lighthouse-core/audits/uses-rel-preload.js]: `isCritical` should really be on `NetworkRequest` or something

**Comment by patrickhulce** [comment]
> What does this mean for data urls that load more stuff from the network? Is the critical chain dropped there anyways or does it properly attribute initiation to the thing containing the data url?

I would be shocked if the initiator is properly attributed to a data URI. Even if it were, I would also be shocked if there were ever a situation in which a chain of critical requests came from a data URI resource. I'm not sure I can recall a time I've seen a critical stylesheet or script served as a data URI instead of just being inlined. 

If we're concerned about this, I'd prefer to build a separate audit that says "why on earth did you ship a critical resource that's not a font or image as a data URI" 😆

**Comment by brendankenny** [comment]
> I would be shocked if the initiator is properly attributed to a data URI. Even if it were, I would also be shocked if there were ever a situation in which a chain of critical requests came from a data URI resource. I'm not sure I can recall a time I've seen a critical stylesheet or script served as a data URI instead of just being inlined.
> 
> If we're concerned about this, I'd prefer to build a separate audit that says "why on earth did you ship a critical resource that's not a font or image as a data URI" 😆

so I couldn't resist and just did an http archive query for this, and there *are* sites that do this :) 

Whether or not including them is a good thing on our part becomes the operative question, I guess. One example:

example critical request chain

https://googlechrome.github.io/lighthouse/viewer/?gist=63e85fa1861ab1cc73248d12ad3d1ab7

**Comment by patrickhulce** [comment]
Color me shocked 😱 

C'mon livingspaces, seriously!?

**Comment by brendankenny** [comment]
one way out of this: this PR is really about fixing data URLs as critical requests themselves, not if they should be part of the critical chain. We discard a critical request from the chain [here if any ancestor isn't critical](https://github.com/GoogleChrome/lighthouse/blob/67460ddf05e73c0485aa48cb66faa58cedfba8b1/lighthouse-core/computed/critical-request-chains.js#L94-L100), we could change that to be if any ancestor isn't critical *and* isn't a data URL, so chains going through a data URL will be preserved, but data URL leaves won't be included as intended by this PR.

**Comment by patrickhulce** [comment]
I personally don't think it's worth fixing just for this visualization, but that solution works for me if y'all do.

**Comment by brendankenny** [comment]
> I personally don't think it's worth fixing just for this visualization, but that solution works for me if y'all do.

worth fixing which? The whole PR or my suggestion? :)

**Comment by patrickhulce** [comment]
> worth fixing which? The whole PR or my suggestion? :)

Oh just not worth fixing the specific issue where data URL descendants disappear, so your suggestion :)

PR still seems fine to me as-is coupled maybe with a new audit to tell livingspaces to stop being so cray cray

**Comment by patrickhulce** [comment]
@connorjclark is this assignment meant for me to pickup the committer portion?

**Comment by connorjclark** [comment]
No, you were the reviewer, but this was before git3po assigned people, so I'm just updated. Next action is still Paul.

**Comment by connorjclark** [comment]
I updated the branch.

**Comment by connorjclark** [comment]
best attempt at a commit title, anyone wanna try improving?

**Comment by connorjclark** [comment]
@paulirish @brendankenny one of you should review this now that I rewrote it.

**Review by paulirish — COMMENTED** [review]
seems good! it wasnt immediately obvious why the L121 check only affected leaf nodes but it made sense after a bit.

nice. 

lgtm (the UI wont let me approve)

**Comment by connorjclark** [comment]
> seems good! it wasnt immediately obvious why the L121 check only affected leaf nodes but it made sense after a bit.

what's funny is that it still doesn't make sense to me, but hard to argue with the tests.

**Pull request merged**